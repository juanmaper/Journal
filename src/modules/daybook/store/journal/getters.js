
// export const myGetter = ( state ) => {
//  return state.something
// }

export const getEntriesByTerm = ( state ) => ( term = '' ) => {
  
  if ( term.length === 0 ) return state.entries 

  return state.entries.filter( entry => entry.text.toLowerCase().includes( term.toLocaleLowerCase() ))
}

export const getEntryById = ( state ) => ( id = '' ) => {

  /* If we return the following sentence directly, then it will be a reference to the state object
  and any modification to it will affect the state. Only mutations should be able to modify it, so 
  instead a destructured object is returned so it has its own space in memory */
  const entry = state.entries.find( entry => entry.id === id)

  if ( !entry ) return

  return { ...entry }

}