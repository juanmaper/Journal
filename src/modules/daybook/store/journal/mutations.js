
// export const myMutation = ( state ) => {

// }

export const setEntries = ( state, entries ) => {

  state.entries = [ ...state.entries, ...entries ]
  state.isLoading = false
}

export const updateEntry = ( state, entry ) => {
  
  const index = state.entries.findIndex( element => element.id == entry.id )
  state.entries[index] = entry
}

export const addEntry = ( state, entry ) => {
  state.entries.unshift( entry )
}

export const deleteEntry = ( state, id ) => {
  state.entries = state.entries.filter( element => element.id !== id )
}