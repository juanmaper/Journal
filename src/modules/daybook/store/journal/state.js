
export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(),
      date: new Date().toDateString(),
      text: 'Dolor tempor tempor aute minim cillum. Nostrud do exercitation ad laboris sint. Commodo proident enim amet nulla velit esse veniam eu aliquip tempor dolore ullamco labore. Consectetur consectetur minim amet sint commodo velit incididunt est ex eiusmod ipsum cupidatat. Officia velit enim mollit amet id fugiat quis laborum Lorem minim ipsum.',
      picture: null
    },
    {
      id: new Date().getTime() + 1000,
      date: new Date().toDateString(),
      text: 'Non dolore dolor ad elit eiusmod Lorem laborum commodo. Mollit ipsum officia non dolor ad Lorem in proident eu. Reprehenderit duis excepteur in fugiat do voluptate aliqua elit consectetur sit ipsum. Dolor fugiat nulla amet incididunt ex incididunt id magna nulla labore non Lorem sint adipisicing. Dolor magna incididunt occaecat aliquip sunt laborum fugiat culpa. Qui culpa consequat culpa minim aliqua consequat mollit.',
      picture: null
    },
    {
      id: new Date().getTime() + 2000,
      date: new Date().toDateString(),
      text: 'Officia fugiat amet duis consequat. Ad ut tempor eu mollit dolore et nulla id. Ad veniam in laboris elit ut quis deserunt labore fugiat. Mollit anim magna exercitation laboris. Dolore do qui consectetur dolore magna amet dolore aliqua.',
      picture: null
    }
  ]
})