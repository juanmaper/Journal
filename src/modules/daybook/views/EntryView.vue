<template>
  <div class="entry-title d-flex justify-content-between p-2">

    <div>
      <span class="text-success fs-3 fw-bold">{{ day }}</span>
      <span class="mx-1 fs-3">{{ month }}</span>
      <span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
    </div>

    <div>
      <input type="file" 
            @change="onSelectedImage"
            ref="imageSelector"
            v-show="false"
            accept="image/png, image/jpeg" >

      <button 
        v-if="entry.id"
        class="btn btn-danger mx-2"
        @click="onDeleteEntry">
          Delete
        <i class="fa fa-trash-alt"></i>
      </button>

      <button class="btn btn-primary" @click="onSelectImage">
        Upload photo
        <i class="fa fa-upload"></i>
      </button>
    </div>
  </div>

  <hr>
  <div class="d-flex flex-column px-3 h-75">
    <textarea 
      v-model="entry.text"
      placeholder="What happened today?">
    </textarea>
  </div>

  <Fab icon="fa-save" @on:click="saveEntry" />

  <img v-if="localImage"
       :src="localImage" 
       alt="Entry picture"
       class="img-thumbnail">
</template>

<script>
import { defineAsyncComponent } from '@vue/runtime-core'
import { mapActions, mapGetters } from 'vuex'
import Swal from 'sweetalert2'

import getDayMonthYear from "../helpers/getDayMonthYear"

export default {
  props: {
    id: {
      type: String,
      required: true
    }
  },

  data() {
    return {
      entry: null,
      localImage: null,
      file: null
    }
  },

  components: {
    Fab: defineAsyncComponent( () => import('../components/Fab.vue'))
  },

  methods: {
    ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry']),
    loadEntry() {
      let entry 

      if ( this.id === 'new') {
        entry = {
          text: '',
          date: new Date().getTime()
        }
      } else {
          entry = this.getEntryById( this.id )
          if ( !entry ) return this.$router.push({ name: 'no-entry' })  
      }

      this.entry = entry
    },
    async saveEntry() {

      new Swal({
        title: 'Saving...',
        allowOutsideClick: false
      })

      Swal.showLoading()

      if ( this.entry.id ) {
        // Update entry
        await this.updateEntry(this.entry)
      } else {
        //Create entry
        const newEntryId = await this.createEntry( this.entry )
        this.$router.push({ name: 'entry', params: { id: newEntryId } })
      }

      Swal.fire('Saved', 'Entry registered successfully', 'success')
    },
    async onDeleteEntry() {

      const { isConfirmed } = await Swal.fire({
        title: 'Are you sure?',
        text: "You can't undo this action",
        showDenyButton: true,
        confirmButtonText: 'Yes, I am sure'
      })

      if ( !isConfirmed ) return

      new Swal({
        title: 'Deleting...',
        allowOutsideClick: false
      })

      Swal.showLoading()

      await this.deleteEntry( this.entry.id )
      this.$router.push({ name: 'no-entry'})

      Swal.fire('Deleted', '', 'success')
    },

    onSelectedImage( event ) {
      const file = event.target.files[0]

      if ( !file ) {
        this.localImage = null
        this.file = null
        return
      }

      this.file = file

      const fr = new FileReader()
      fr.onload = () => this.localImage = fr.result
      fr.readAsDataURL( file )

      console.log(this.localImage);
      console.log('-----------------------');
      console.log(this.file);
    },

    onSelectImage() {
      this.$refs.imageSelector.click()
    }
  },

  computed: {
    ...mapGetters('journal', ['getEntryById']),
    day() {
      const { day } = getDayMonthYear( this.entry.date )
      return day
    },
    month() {
      const { month } = getDayMonthYear( this.entry.date )
      return month
    },
    yearDay() {
      const { yearDay } = getDayMonthYear( this.entry.date )
      return yearDay
    }
  },

  created() {
    this.loadEntry()
  },

  watch: {
    id() {
      this.loadEntry()
    }
  }
}
</script>

<style lang="scss" scoped>

textarea {
  font-size: 20px;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
}
</style>