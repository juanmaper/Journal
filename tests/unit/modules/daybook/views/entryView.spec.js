import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'

import journal from '@/modules/daybook/store/journal'
import { journalState } from '../../../mocks-data/test-journal-state'

import EntryView from '@/modules/daybook/views/EntryView'

const createVuexStore = ( initialState ) => 
  createStore({
    modules: {
      journal: {
        ...journal,
        state: structuredClone( initialState )
      }
    }
  })

describe('EntryView tests', () => {

  const store = createVuexStore( journalState )
  const mockRouter = {
    push: jest.fn()
  }

  let wrapper

  beforeEach( () => {

    jest.clearAllMocks()

    wrapper = shallowMount( EntryView, {
      props: {
        id: '-N4Wv4XVronCdn1koj1B'
      },
      global: {
        mocks: {
          $router: mockRouter
        },
        plugins: [ store ]
      }
    })
  })
  
  test('should redirect the user if the id does not exist', () => {
    
    shallowMount( EntryView, {
      props: {
        id: 'This ID does not exist in the STORE'
      },
      global: {
        mocks: {
          $router: mockRouter
        },
        plugins: [ store ]
      }
    })

    expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })

  })

  test('should show the entry correctly', () => {
    
    expect( wrapper.html() ).toMatchSnapshot()
    expect( mockRouter.push ).not.toHaveBeenCalled()

  })

})    