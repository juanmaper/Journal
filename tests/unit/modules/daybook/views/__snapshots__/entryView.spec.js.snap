// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryView tests should show the entry correctly 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">14</span><span class="mx-1 fs-3">Junio</span><span class="mx-2 fs-4 fw-light">2022, Martes</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload photo <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="What happened today?">
      </textarea></div>
<img src="https://res.cloudinary.com/ds3yf4btu/image/upload/v1655216917/iczhdkowquouhqjer3zp.png" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
